### ISSUE TITLE
Add replay for GitLab.com API calls that result in 500 Internal Server Error

### ISSUE THREAD

**Issue by rossbeehler** [gitlabform/gitlabform#531]
We get an error like this intermittently, and was hoping retry logic could be added around this request:

```
(3247/3926) Processing project: path/to/project
Warning: Error occurred while processing project path/to/project, exception:
Request url='[https://gitlab.com/api/v4/projects/path%2Fto%2Fproject/protected_branches?name=main'](https://gitlab.com/api/v4/projects/path%2Fto%2Fproject/protected_branches?name=main%27), method=POST, json='{"allow_force_push": true, "merge_access_level": 40, "push_access_level": 40}' failed - expected code(s) [200, 201], got code 500 & body: '{"message":"500 Internal Server Error"}' 
```
This particular example is for the `protected_branches` resource, but there might be others that could benefit from additional retry.

Perhaps for 500 requests, a small wait could be added between retries?

**Comment by rossbeehler** [comment]
Is there a way to see when a retry is executed?  I think you have these 3 retries shared across all API calls for an entire gitlabform run, correct?  Since we have ~4000 projects, I can see those getting used up quickly over the 3.5 hours it currently takes to run.

**Comment by gdubicki** [comment]
We have 1 original request + 3 retries = 4 tries _in total for each request, not for the entire run_.
Also we have a retry of the whole section added in https://github.com/gitlabform/gitlabform/pull/497.
But the error you have shared here is a HTTP 500 in response to a POST and these are not retried as of now, see https://github.com/gitlabform/gitlabform/issues/492#issuecomment-1435657246 for more info why.
So we definitely can think of adding retries here too, just have to be smart about it as POSTs are not idempotent. Maybe we should cover this case too in https://github.com/gitlabform/gitlabform/blob/7d77a253f88422e954485aea3c61ecbd6b90a75f/gitlabform/processors/abstract_processor.py#L103-L125 ? 🤔

**Comment by gdubicki** [comment]
> Is there a way to see when a retry is executed?

Yes, with a debug output. But I understand that it's not practical with your scale. To make the retries more exposed we would have to have a custom subclass of the standard Retry that would print more info with `cli_ui`, something like this:

```python
original_total = 3

class VerboseRetry(Retry):
    def __init__(self, *args, **kwargs):
        super(VerboseRetry, self).__init__(*args, **kwargs)

    def increment(
            self,
            *args, **kwargs
    ):
        to_return = super(VerboseRetry, self).increment(*args, **kwargs)
        if not to_return.is_exhausted():
            cli_ui.info(f"Retry {original_total-to_return.total} of {original_total}...")
        return to_return
```

I'll think about that.

**Comment by hiaik** [comment]
Just in case, for original issue reported

```
(X/Y) Processing project: path/to/project
Warning: Error occurred while processing project path/to/project, exception:
Request url='[https://gitlab.com/api/v4/projects/path%2Fto%2Fproject/protected_branches?name=main'](https://gitlab.com/api/v4/projects/path%2Fto%2Fproject/protected_branches?name=main%27), method=POST, json='{"allow_force_push": true, "merge_access_level": 40, "push_access_level": 40}' failed - expected code(s) [200, 201], got code 500 & body: '{"message":"500 Internal Server Error"}' 
```

gitlab has https://gitlab.com/gitlab-org/gitlab/-/issues/463017

**Comment by hiaik** [comment]
Hi @gdubicki and @amimas 

Will you accept MR with changes for POST requests (retry) or it is not something you see as a feature for a product?

**Comment by hiaik** [comment]
> I would assume that having feature like retry per project for POST would be useful. For example:
> By default no retries for POST requests per project
> If gitlab form configuration has "project" -> retryPOST: N, then POST requests should be retried for 409, 500 and 502 error codes N times using default timeout between attempts.

After more thinking here... Maybe it will be better to consider retry in terms of project and not just POST request.

For example:
1. By default, no retries
2. If config has "project" -> "retry: N" and that "project" processing failed, it should be retried from the beginning.
3. Maybe we can consider retrying all failed project at the end of the job (when we have a summary on what failed).
